File: pt5/1177.can-make-palindrome-from-substring.java
import java.util.*;

class Solution {
    public List<Boolean> canMakePaliQueries(String s, int[][] queries) {
        int[] prefix = new int[s.length() + 1];
        int mask = 0;

        for (int i = 0; i < s.length(); i++) {
            mask ^= 1 << (s.charAt(i) - 'a');
            prefix[i + 1] = mask;
        }

        List<Boolean> answer = new ArrayList<>(queries.length);
        for (int[] query : queries) {
            int left = query[0];
            int right = query[1];
            int k = query[2];
            int oddMask = prefix[right + 1] ^ prefix[left];
            int oddCount = Integer.bitCount(oddMask);
            answer.add(oddCount / 2 <= k);
        }
        return answer;
    }
}

File: pt5/1177.can-make-palindrome-from-substring.py
from __future__ import annotations

from typing import List


class Solution:
    def canMakePaliQueries(self, s: str, queries: List[List[int]]) -> List[bool]:
        prefix = [0]
        mask = 0

        for char in s:
            mask ^= 1 << (ord(char) - ord("a"))
            prefix.append(mask)

        answer = []
        for left, right, k in queries:
            odd_mask = prefix[right + 1] ^ prefix[left]
            odd_count = odd_mask.bit_count()
            answer.append(odd_count // 2 <= k)

        return answer

